users/views_test.py

```diff
@@ -118,3 +123,89 @@ def test_countries_states_view(client):
     assert len(countries.get("FR").get("states")) == 0
     assert countries.get("US").get("name") == "United States"
     assert countries.get("TW").get("name") == "Taiwan"
+
+
+def test_create_email_change_request_invalid_password(user_drf_client, user):
+    """Test that invalid password is returned"""
+    resp = user_drf_client.post(
+        "/api/change-emails/",
+        data={
+            "new_email": "abc@example.com",
+            "password": user.password,
+            "old_password": "abc",
+        },
+    )
+    assert resp.status_code == status.HTTP_400_BAD_REQUEST
+
+
+def test_create_email_change_request_existing_email(user_drf_client, user):
+    """Test that create change email request gives validation error for existing user email"""
+    new_user = UserFactory.create()
+    user_password = user.password
+    user.set_password(user.password)
+    user.save()
+    resp = user_drf_client.post(
+        "/api/change-emails/",
+        data={"new_email": new_user.email, "password": user_password},
+    )
+    assert resp.status_code == status.HTTP_400_BAD_REQUEST
+
+
+def test_create_email_change_request_valid_email(user_drf_client, user, mocker):
+    """Test that change request is created"""
+    user_password = user.password
+    user.set_password(user.password)
+    user.save()
+
+    mocker.patch("courseware.tasks.change_edx_user_email_async", return_value=None)
+    mocker.patch("courseware.tasks.api.update_edx_user_email")
+    mock_email = mocker.patch("mail.verification_api.send_verify_email_change_email")
+    resp = user_drf_client.post(
+        "/api/change-emails/",
+        data={"new_email": "abc@example.com", "password": user_password},
+    )
+
+    assert resp.status_code == status.HTTP_201_CREATED
+
+    code = mock_email.call_args[0][1].code
+    assert code
+
+    old_email = user.email
+    resp = user_drf_client.patch(
+        "/api/change-emails/{}/".format(code), data={"confirmed": True}
+    )
+    assert not UserSocialAuth.objects.filter(uid=old_email, user=user).exists()
+    assert resp.status_code == status.HTTP_200_OK
+    user.refresh_from_db()
+    assert user.email == "abc@example.com"
+
+
+def test_create_email_change_request_expired_code(user_drf_client, user):
+    """Check for expired code for Email Change Request"""
+    change_request = ChangeEmailRequest.objects.create(
+        user=user, new_email="abc@example.com", expires_on=now_in_utc()
```

Comment:
This could end up being a flaky test if it runs quick enough, making it expire a few seconds in the past would be safer.

Reply:
ok